### ISSUE TITLE
"Error: Found an unshardable source dataset" when using Mirrored Strategy with train on batch

### ISSUE THREAD

**Issue by houseofai** [tensorflow/tensorflow#45157]
**System information**
- Have I written custom code: Custom
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- TensorFlow installed from (source or binary): Tested from compiled source and binary
- TensorFlow version (use command below): 2.4.0rc2
- Python version: 3.9/3.8/3.7
- Bazel version (if compiling from source): 3.7.0
- GCC/Compiler version (if compiling from source): 9.3.0/7
- CUDA/cuDNN version: 11.1/10.1
- GPU model and memory: RTX 3090/ RTX 2080

**Describe the current behavior**
When using MirroredStrategy with Keras, it produces a lot of outputs (see below) as described in this issue #42146. For each batch, it keeps printing the same logs with the error `Error: Found an unshardable source dataset`
I tried to disable the `tf.data.experimental.AutoShardPolicy.OFF` on `train_dataset` or to set it to DATA, but it seems to not have any effect (probable because the `for loop` unpack the train dataset)
It works without any problem on TF 2.3.1

**Describe the expected behavior**
I expect to have the same behavior as on TF 2.3.1: no extensive printed log, no error. Have it works with the default shard policy.
Or at least, to find a way to apply the shard policy on the two subset train_images, train_labels

**Standalone code to reproduce the issue**
```
from tensorflow.keras import datasets, layers, models
import tensorflow as tf

options = tf.data.Options()
options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
(train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()
train_images, test_images = train_images / 255.0, test_images / 255.0
train_dataset = tf.data.Dataset.from_tensor_slices((train_images, train_labels))
train_dataset = train_dataset.with_options(options)
train_dataset = train_dataset.batch(32)

strategy = tf.distribute.MirroredStrategy()
with strategy.scope():
    model = models.Sequential()
    model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.MaxPooling2D((2, 2)))
    model.add(layers.Conv2D(64, (3, 3), activation='relu'))
    model.add(layers.Flatten())
    model.add(layers.Dense(64, activation='relu'))
    model.add(layers.Dense(10))

    model.compile(optimizer='adam',
                  loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
                  metrics=['accuracy'])


for train_images, train_labels in train_dataset:
    history = model.train_on_batch(train_images, train_labels)
```

**Other info / logs** 
```
2020-11-24 17:17:49.116426: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2020-11-24 17:17:49.885169: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2020-11-24 17:17:49.885621: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcuda.so.1
2020-11-24 17:17:49.917150: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1727] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce RTX 3090 computeCapability: 8.6
coreClock: 1.695GHz coreCount: 82 deviceMemorySize: 23.69GiB deviceMemoryBandwidth: 871.81GiB/s
2020-11-24 17:17:49.917174: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2020-11-24 17:17:49.918707: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2020-11-24 17:17:49.918727: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2020-11-24 17:17:49.919263: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2020-11-24 17:17:49.919382: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2020-11-24 17:17:49.920901: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.11
2020-11-24 17:17:49.921285: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2020-11-24 17:17:49.921350: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2020-11-24 17:17:49.922684: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1869] Adding visible gpu devices: 0
2020-11-24 17:17:49.923076: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE3 SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2020-11-24 17:17:49.924740: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set
2020-11-24 17:17:49.925430: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1727] Found device 0 with properties: 
pciBusID: 0000:01:00.0 name: GeForce RTX 3090 computeCapability: 8.6
coreClock: 1.695GHz coreCount: 82 deviceMemorySize: 23.69GiB deviceMemoryBandwidth: 871.81GiB/s
2020-11-24 17:17:49.925441: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2020-11-24 17:17:49.925448: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2020-11-24 17:17:49.925455: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
2020-11-24 17:17:49.925462: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcufft.so.10
2020-11-24 17:17:49.925468: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcurand.so.10
2020-11-24 17:17:49.925474: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusolver.so.11
2020-11-24 17:17:49.925480: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcusparse.so.11
2020-11-24 17:17:49.925486: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2020-11-24 17:17:49.926781: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1869] Adding visible gpu devices: 0
2020-11-24 17:17:49.926800: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudart.so.11.0
2020-11-24 17:17:50.236484: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1267] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-11-24 17:17:50.236514: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1273]      0 
2020-11-24 17:17:50.236519: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1286] 0:   N 
2020-11-24 17:17:50.238508: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1413] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 22185 MB memory) -> physical GPU (device: 0, name: GeForce RTX 3090, pci bus id: 0000:01:00.0, compute capability: 8.6)
2020-11-24 17:17:51.610137: W tensorflow/core/grappler/optimizers/data/auto_shard.cc:656] In AUTO-mode, and switching to DATA-based sharding, instead of FILE-based sharding as we cannot find appropriate reader dataset op(s) to shard. Error: Found an unshardable source dataset: name: "TensorDataset/_2"
op: "TensorDataset"
input: "Placeholder/_0"
input: "Placeholder/_1"
attr {
  key: "Toutput_types"
  value {
    list {
      type: DT_DOUBLE
      type: DT_UINT8
    }
  }
}
attr {
  key: "output_shapes"
  value {
    list {
      shape {
        dim {
          size: 32
        }
        dim {
          size: 32
        }
        dim {
          size: 32
        }
        dim {
          size: 3
        }
      }
      shape {
        dim {
          size: 32
        }
        dim {
          size: 1
        }
      }
    }
  }
}


```
Thank you!

**Comment by nikitamaia** [comment]
Hi @houseofai, I was able to repro this and I see that the logs look different in 2.3 vs in nightly. I also was able to repro using `model.fit` instead of `model.train_on_batch`.

Note that #42146 was in reference to `MultiWorkerMirroredStrategy`, and the use of an autoshard policy is not necessary when using `MirroredStrategy` since you only have one machine. This can be a bit confusing (especially given the logs you are seeing), so in case you haven't already taken a look at the [Distributed Input guide](https://www.tensorflow.org/tutorials/distribute/keras) I am linking it here.

**Comment by crccw** [comment]
This is a warning, not an error. The warning is expected, as we cannot shard tensor based dataset by FILE.

You can pass the the dataset with options to model.fit() and the option should take effect.

**Comment by houseofai** [comment]
Thanks @nikitamaia 
Should I understand that now the default behavior with MirroredStrategy and train_on_batch is to print those verbose warnings all the way long the training?

**Comment by nikitamaia** [comment]
Yes, the reason the options aren't working is that `model.train_on_batch` wraps the tensors in a dataset again, so the specified options are not taking effect. If you pass the dataset to `model.fit` with the options then you won't see the warning. That being said, since there is no autosharding when you're doing single worker training the autosharding options register as no-op, but the code is shared with MWMS so you see the warning regardless. I think ideally the verbose warnings shouldn't be there and in the future we can improve this, but for now the training is working as expected.

**Comment by houseofai** [comment]
Thanks a lot @nikitamaia for the detailed answers

**Comment by Flamefire** [comment]
@crccw Why was this closed? As @nikitamaia explained using options to steer the sharding of a dataset passed to `model.train_on_batch` is not working due to some internals of that function. The proposed "solution" is to not use this function which is hardly a solution.

So may I rephrase the issue/bug report as "Dataset options ignored for dataset passed to `model.train_on_batch`" with OPs code  as a reproducer. And although sharding is not required for MirroredStrategy this is still a valid bug in TF 2.4: a) The option is clearly ignored and b) the same happens for MultiWorkerMirroredStrategy

**Comment by motionlife** [comment]
Can't believe they close this problem, this is atrocious

**Comment by nikitamaia** [comment]
Hi all, it seems that my explanation resulted in more confusion. 

The solution was not to avoid `train_on_batch`, but instead to explain that `train_on_batch` will result in some warnings that you can ignore. Ideally, we would silence these warnings, but for now they are just warnings that don't actually affect the training process.

The sharding options set on the original dataset don't apply to `train_on_batch` because `train_on_batch` takes tensors as inputs and creates a new dataset internally. This is intended Keras behavior.

Because the options are not taking effect, you are unable to explicitly set the autoshardpolicy. Thus, the autoshardpolicy is set to AUTO as indicated in the docs:

```
AUTO: This is the default option which means an attempt will be made to shard by FILE. 
The attempt to shard by FILE fails if a file-based dataset is not detected. tf.distribute will then fall back to sharding by DATA. 
```

So first the policy tries to shard by FILE, but it cannot since `train_on_batch` takes tensors as input. So then the DATA policy is used.`Error: Found an unshardable source dataset` is indicating that the input could not be sharded by FILE, and will instead be sharded by DATA.

To summarize, it seems the main point of confusion here is the warning `Error: Found an unshardable source dataset` which in this case is telling us that with `model.train_on_batch` you cannot shard by FILE. However, the phrasing of the warning is confusing/misleading. 

A fix in this case either means writing a better warning message, or from the Keras side setting the autoshardpolicy to DATA when `train_on_batch` is used. In the end the same effect occurs, but just without the warning message. 

Hopefully this clears things up!

**Comment by santiagoRuizSchiphol** [comment]
Hi all, please be aware that you also have this warring message not only at training time but also when you are just predicting and you want to leverage with `MirroredStrategy` to increase the batch size.
